### DISCUSSION TITLE
Bug with CSRF Token with HMAC

### DISCUSSION THREAD

**Discussion by yazalulloa** [quarkusio/quarkus#36821] [Q&A]
Hi, I'm trying to build a web application with HTMX, simple js library to use AJAX requests, for POST, DELETE and PUT I understand I need to pass the CSRF token either as form param or as a header, I'm finding that after I refresh the page the CSRF token that gets injected in the HTML is the previous cookie token, every HTTP request except for GET fails. This only happens if I have the property quarkus.csrf-reactive.token-signature-key.

Is there some missing configuration I don't have or is it a bug?

Here is the code
https://github.com/yazalulloa/quarkus-csrf-signature-bug

First load:
![image](https://github.com/quarkusio/quarkus/assets/33322056/7a3da45b-25cd-4a62-877c-fe734621c494)
![image](https://github.com/quarkusio/quarkus/assets/33322056/4cef6e8f-c550-4d7f-96f9-3b39332cd4eb)

Second load:
![image](https://github.com/quarkusio/quarkus/assets/33322056/7cc926d1-40b5-4e13-af3b-49600616fbdc)
![image](https://github.com/quarkusio/quarkus/assets/33322056/9baf9876-f5c8-4c09-b629-65c2299e6fd4)

**Comment by sberyozkin** [comment]
@yazalulloa thanks for the investigation, 

>I'm finding that after I refresh the page the CSRF token that gets injected in the HTML is the previous cookie token, every HTTP request except for GET fails. This only happens if I have the property quarkus.csrf-reactive.token-signature-key.

It is a bug, when the signature is required, the cookie value is a signed csrf token, so, after the refresh, if it is returned as a CSRF token (which is fine without the signature), then it fails because when later the the csrf token is verified again, it fails because the already signed value is now double signed and therefore fails to match the cookie value.

Please move this discussion to the issue as a bug, thanks

**Reply by yazalulloa** [reply]
Done https://github.com/quarkusio/quarkus/issues/36848
